qa/os/src/test/java/org/elasticsearch/packaging/util/Docker.java

```diff
@@ -0,0 +1,382 @@
+     */
+    public static Installation runContainer(Distribution distribution, Path configPath, Map<String,String> envVars) throws Exception {
+        removeContainer();
+
+        final List<String> args = new ArrayList<>();
+
+        args.add("docker run");
+
+        // Remove the container once it exits
+        args.add("--rm");
+
+        // Run the container in the background
+        args.add("--detach");
+
+        envVars.forEach((key, value) -> args.add("--env " + key + "=\"" + value + "\""));
+
+        // The container won't run without configuring discovery
+        args.add("--env discovery.type=single-node");
+
+        // Map ports in the container to the host, so that we can send requests
+        args.add("--publish 9200:9200");
+        args.add("--publish 9300:9300");
+
+        if (configPath != null) {
+            // Bind-mount the config dir, if specified
+            args.add("--volume \"" + configPath + ":/usr/share/elasticsearch/config\"");
+        }
+
+        args.add(distribution.flavor.name + ":test");
+
+        final String command = String.join(" ", args);
+        logger.debug("Running command: " + command);
+        containerId = sh.run(command).stdout.trim();
+
+        waitForElasticsearchToStart();
+
+        return Installation.ofContainer();
+    }
+
+    /**
+     * Waits for the Elasticsearch process to start executing in the container.
+     * This is called every time a container is started.
+     */
+    private static void waitForElasticsearchToStart() throws InterruptedException {
+        boolean isElasticsearchRunning = false;
+        int attempt = 0;
+
+        do {
+            String psOutput = dockerShell.run("ps ax").stdout;
+
+            if (psOutput.contains("/usr/share/elasticsearch/jdk/bin/java -X")) {
+                isElasticsearchRunning = true;
+                break;
+            }
+
+            Thread.sleep(1000);
+        } while (attempt++ < 5);
+
+        if (!isElasticsearchRunning) {
+            final String logs = sh.run("docker logs " + containerId).stdout;
+            fail("Elasticsearch container did start successfully.\n\n" + logs);
+        }
+    }
+
+    /**
+     * Waits for the Elasticsearch process to finish booting up. This takes a few seconds, so this action
+     * is not taken every time a container is started, since many tests check the Docker image in
+     * some way, and not Elasticsearch specifically.
+     */
+    public static void waitForElasticsearchToBecomeAvailable(Distribution distribution) throws InterruptedException {
+        int attempt = 0;
+
+        // Give Elasticsearch time to become available. I measured roughly 10s on my laptop
+        Thread.sleep(10 * 1000);
+
+        do {
+            String dockerLogs = sh.run("docker logs " + containerId).stdout;
+
+            if (distribution.isOSS() && dockerLogs.contains("recovered [0] indices into cluster_state")) {
```

Comment:
Could we implement this similarly to other distributions by checking the cluster health endpoint ? 
It's counter intuitive to have a failure in a packaging test when changing a log line.

Reply:
Thanks, I've done as you suggested.